src/Algebra/Lattice/Heyting.hs

```diff
@@ -0,0 +1,346 @@
+  , implies
+  , (<=>)
+  , iff
+  , iff'
+  , toBoolean
+#ifdef EXPORT_PROPERTIES
+    -- * QuickCheck Properties
+    --
+    -- $properties
+  , prop_BoundedMeetSemiLattice
+  , prop_BoundedJoinSemiLattice
+  , prop_DistributiveLattice
+  , prop_HeytingAlgebra
+  , prop_implies
+#endif
+  )
+  where
+
+import Prelude hiding (not)
+import qualified Prelude
+
+import Control.Applicative    (Const (..))
+#ifdef EXPORT_PROPERTIES
+import Data.List              (intersperse)
+#endif
+import Data.Functor.Identity  (Identity (..))
+import Data.Hashable          (Hashable)
+import Data.Proxy             (Proxy (..))
+import Data.Semigroup         (All (..), Any (..), Endo (..), (<>))
+#ifdef EXPORT_PROPERTIES
+import Data.Monoid            (mempty)
+#endif
+import Data.Tagged            (Tagged (..))
+import Data.Universe.Class    (Finite, universe)
+import qualified Data.Map as M
+import qualified Data.Set as S
+import qualified Data.HashMap.Lazy as HM
+import qualified Data.HashSet      as HS
+
+#ifdef EXPORT_PROPERTIES
+import Algebra.Lattice ( BoundedJoinSemiLattice (..)
+                       , BoundedMeetSemiLattice (..)
+                       , Lattice
+                       , Join (..)
+                       )
+#endif
+import Algebra.Lattice ( BoundedLattice
+                       , Meet (..)
+                       , bottom
+                       , top
+                       , (/\)
+                       , (\/)
+                       )
+import Algebra.Lattice.Dropped    (Dropped    ( ..))
+import Algebra.Lattice.Lifted     (Lifted     ( ..))
+import Algebra.Lattice.Levitated  (Levitated)
+import qualified Algebra.Lattice.Levitated as L
+import Algebra.Lattice.Ordered    (Ordered    ( ..))
+import Algebra.PartialOrd         (leq)
+#ifdef EXPORT_PROPERTIES
+import Test.QuickCheck            (Blind (..), Property, counterexample, (.&&.), (===))
+import qualified Test.QuickCheck as QC
+#endif
+
+-- |
+-- Heyting algebra is a bounded semi-lattice with implication which should
+-- satisfy the following law:
+--
+-- prop> x ∧ a ≤ b ⇔ x ≤ (a ⇒ b)
+--
+-- We also require that a Heyting algebra is a distributive lattice, which
+-- means any of the two equialent conditions holds:
+--
+-- prop> a ∧ (b ∨ c) = a ∧ b ∨ a ∧ c
+-- prop> a ∨ (b ∧ c) = (a ∨ b) ∧ (a ∨ c)
+--
+-- This means @a ⇒ b@ is an
+-- [exponential object](https://ncatlab.org/nlab/show/exponential%2Bobject),
+-- which makes any Heyting algebra
+-- a [cartesian
+-- closed category](https://ncatlab.org/nlab/show/cartesian%2Bclosed%2Bcategory).
+-- This means that Curry isomorphism holds (which takes a form of an actual
+-- equality):
+--
+-- prop> a ⇒ (b ⇒ c) = (a ∧ b) ⇒ c
+--
+-- Some other useful properties of Heyting algebras:
+-- 
+-- prop> (a ⇒ b) ∧ a ≤ b
+-- prop> b ≤ a ⇒ a ∧ b
+-- prop> a ≤ b  iff a ⇒ b = ⊤
+-- prop> b ≤ b' then a ⇒ b ≤ a ⇒ b'
+-- prop> a'≤ a  then a' ⇒ b ≤ a ⇒ b
+-- prop> not (a ∧ b) = not (a ∨ b)
+-- prop> not (a ∨ b) = not a ∧ not b
```

Comment:
note to self: https://en.wikipedia.org/wiki/Heyting_algebra#The_De_Morgan_laws_in_a_Heyting_algebra

Add an counterexample.